### PR TITLE
Big rewrite

### PR DESCRIPTION
**Pull request by phadej**
I tried to comment code, I'll be happy to clarify.

### CODE REVIEW

**Comment by mpickering** [comment]
Can we review and merge this @bgamari ?

File: cbits/eventlog_socket.c [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+
+/*********************************************************************************
+ * Public interface
+ *********************************************************************************/
+
+void eventlog_socket_wait(void)
+{
+  pthread_mutex_lock(&mutex);
+  while (client_fd == -1) {
+    assert(pthread_cond_wait(&new_conn_cond, &mutex) == 0);
```
Comment (tstat) [inline cbits/eventlog_socket.c:489]: Don't we want to call `pthread_cond_wait` outside of an assert so that it gets executed even if NDEBUG is set?

**Review by bgamari — COMMENTED** [review]
Goods quite good. Thanks, @phadej

**Comment by bgamari** [comment]
I have gone ahead and merged this but it appears that there is something wrong here; a simple "hello world" testcase compiled with 8.10.7 (from #4) seems to result in an eventlog without a header.

@phadej, were you able to test this?

**Comment by bgamari** [comment]
I did a bit of digging here. The problem appears to be that the `endEventLogging`, which `listen_iteration` calls, pushes a block marker to the `EventBuf`.  This then appears before the eventlog header (produced by `startEventLogging`) in the output stream. I fixed this in https://gitlab.haskell.org/ghc/ghc/-/commit/8188adf0f1a0482c269d1eb6350dd91dddc9ed29. Unfortunately, that commit is only in GHC 9.2 and later.

**Comment by bgamari** [comment]
Indeed I can confirm that things work as expected in GHC 9.2.

**Comment by phadej** [comment]
@bgamari I'm not working on this anymore. Please take over.

**Pull request merged**